gp_emulator/GaussianProcess.py

```diff
@@ -221,35 +223,113 @@ def predict ( self, testing, do_unc=True ):
 	do_unc: flag, optional
 		Calculate the uncertainty (if you don't set this flag, it
 		can shave a few us"""
-
+       
 
         ( nn, D ) = testing.shape
         assert D == self.D
-        
         expX = np.exp ( self.theta )
         
         a = dist.cdist ( np.sqrt(expX[:(self.D)])*self.inputs, \
             np.sqrt(expX[:(self.D)])*testing, 'sqeuclidean')
-        
         a = expX[self.D]*np.exp(-0.5*a)
         b = expX[self.D]
         
         mu = np.dot( a.T, self.invQt)
+
         if do_unc:
-	    var = b - np.sum (  a * np.dot(self.invQ,a), axis=0)
+	       var = b - np.sum (  a * np.dot(self.invQ,a), axis=0)
         # Derivative and partial derivatives of the function
         deriv = np.zeros ( ( nn, self.D ) )
 
         for d in xrange ( self.D ):
             aa = self.inputs[:,d].flatten()[None,:] - testing[:,d].flatten()[:,None]
             c = a*aa.T
-
             deriv[:, d] = expX[d]*np.dot(c.T, self.invQt)
         if do_unc:
             return mu, var, deriv
         else:
 	    return mu, deriv
         
+        
+    def gpu_predict ( self, testing, prec, threashold):# self, testing, do_unc=True):
+        import _gpu_predict
+        ( nn, D ) = testing.shape
+        assert D == self.D
+        expX = np.exp ( self.theta )
+        
+        N=testing.shape[0]
+        M=self.inputs.shape[0]
+        theta_size=self.theta.size
+
+        nblocks = np.int(np.ceil( np.float32( nn ) / threashold ))
+        block_size = np.int(np.ceil( nn / nblocks ))
+        ind_start = range( np.int(0), np.int(nn), np.int(block_size) )
+        ind_end = np.append( ind_start[1:], nn )
+        nblocks = len(ind_start)
+        
+        # the last two block should have equal size
+        if nblocks > 1:
+            last_two_block_size = ( ind_end[ nblocks - 1 ] - ind_start[ nblocks - 2 ] ) / 2
+            ind_end[ nblocks - 2 ] = ind_start[ nblocks - 2 ] + last_two_block_size
+            ind_start[ nblocks - 1 ] = ind_end[ nblocks - 2 ]
+        
+        assert np.all(ind_end - ind_start > 500 )
+        assert np.all(ind_end - ind_start < threashold * 1.5 )
+            
+
+
+        #stretch the matrices to vector to feed _predict_wrap()
+        inputs = self.inputs.reshape(self.inputs.shape[0] * self.inputs.shape[1])
+        invQt = self.invQt
+        invQ =  self.invQ.reshape(self.invQ.shape[0] * self.invQ.shape[1])
+
+        if prec == "float32":
```

Comment:
1. Consider defining precision to be a numpy datatype object rather than a string. With a default `prec=np.float64`. Then you could do: `inputs=precision(inputs)`